File: banco_problemas_aplicados/nivel_intermedio/soluciones/java/Problema29.java
/*
 * Problema 29 - Suma de Matriz de Nodos por Filas
 * En Java una matriz es un arreglo de arreglos (double[][]) que conoce sus dimensiones:
 * mat.length filas, mat[i].length columnas. (Ver teoria/02_matrices_2d.md)
 */
import java.util.Scanner;
import java.util.Locale;

public class Problema29 {

    static double[] sumaFilas(double[][] mat) {
        double[] res = new double[mat.length];
        for (int i = 0; i < mat.length; i++) {
            double suma = 0.0;
            for (int j = 0; j < mat[i].length; j++) {
                suma = suma + mat[i][j];
            }
            res[i] = suma;
        }
        return res;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Numero de filas (N): ");
        int n = sc.nextInt();
        System.out.print("Numero de columnas (M): ");
        int m = sc.nextInt();

        double[][] mat = new double[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                System.out.print("mat[" + i + "][" + j + "]: ");
                mat[i][j] = sc.nextDouble();
            }
        }

        double[] res = sumaFilas(mat);

        System.out.println("Suma por filas:");
        for (int i = 0; i < n; i++) {
            System.out.printf(Locale.US, "  fila %d = %.2f%n", i, res[i]);
        }
        sc.close();
    }
}

File: banco_problemas_aplicados/nivel_intermedio/soluciones/cpp/problema29.cpp
/*
 * Problema 29 - Suma de Matriz de Nodos por Filas
 * En C++ una matriz dinamica se representa comodamente con vector<vector<double>>,
 * que lleva sus propias dimensiones (mat.size() filas, mat[i].size() columnas).
 * (Ver teoria/02_matrices_2d.md)
 */
#include <iostream>
#include <iomanip>
#include <vector>
using namespace std;

vector<double> sumaFilas(const vector<vector<double>>& mat) {
    vector<double> res(mat.size());
    for (size_t i = 0; i < mat.size(); i++) {
        double suma = 0.0;
        for (size_t j = 0; j < mat[i].size(); j++) {
            suma = suma + mat[i][j];
        }
        res[i] = suma;
    }
    return res;
}

int main() {
    int n, m;
    cout << "Numero de filas (N): ";
    cin >> n;
    cout << "Numero de columnas (M): ";
    cin >> m;

    vector<vector<double>> mat(n, vector<double>(m));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cout << "mat[" << i << "][" << j << "]: ";
            cin >> mat[i][j];
        }
    }

    vector<double> res = sumaFilas(mat);

    cout << fixed << setprecision(2);
    cout << "Suma por filas:" << endl;
    for (int i = 0; i < n; i++) {
        cout << "  fila " << i << " = " << res[i] << endl;
    }
    return 0;
}
